CrashReportAnalyser that tries to identify the mod that crashed the game, implements #9717

thedead2 opened the pull request:
This pull request implements one class that tries to automatically identify the mod that caused a crash and append this information to the crash report.
I think this would make it far easier for everyone to identify what caused the game to crash, be it a normal user, a mod developer or a (mod) pack creator.

The CrashReportAnalyser class checks the stacktrace of the exception of a crash report for any occurrence of the base package name (e.g. "net.minecraftforge", "net.minecraft", "de.thedead2", etc.) of a mod and if it identifies the mod, adds the position where the error occurred (the stacktrace element or the mixin class and its target(s)) and the IModInfo to the suspected mods. 
It then iterates over the suspectedMods and appends for each, the mod name, the mod id, its version, the position of the exception and if available the issues link for the mod to the crash report.

Two examples for how the crash report would look like with the suspected mods added:
https://gist.github.com/thedead2/af7f23bba630586755903866820e8f2f
https://gist.github.com/thedead2/e4bcddc8057ab337894e9d55a39b72f7

I tested the functionality with the following test mod:
https://github.com/thedead2/testmod

I tried to explain everything in detail in form of java doc comments for each method inside the CrashReportAnalyser class. 
I also tried to keep the changes to the minecraft code base as small as possible only adding three lines in the CrashReport class.

Hope this is not too intrusive and fits your liking.

thedead2 committed:
Implements #9717

PaintNinja commented:
Sounds like a good idea, but I'm a bit weary of false-positives - for example, if another mod maliciously coremods mine to make my mod crash, do I get the blame or them? What about if a library mod crashes because I accidentally passed null to one of its methods?

Listing multiple potential causes in the suspected mods could help with this.

thedead2 commented:
Actually I haven't tested this but the CrashReportAnalyser class adds a list of suspected mods to the crash report including all mods that are visible in the stacktrace. So if another mod would invoke a method of your mod it would identify both mods as suspected. This would also be the case if the other mod would try to use mixin to edit the behaviour of your mods code.
So false-positives can occur but I can't think of a way to prevent this at least in the way I have implemented it. 
It's still a best guess of what caused the crash.

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,280 @@
+     * Checks the given stacktrace for any occurrence of the base package names cached in the basePackageModCache,
+     * including mixin classes.
+     * **/
+    private static void scanStacktrace(StackTraceElement[] stackTrace){
+        for(StackTraceElement stackTraceElement : stackTrace){
```
spaces after keywords and between brackets please
```suggestion
    private static void scanStacktrace(StackTraceElement[] stackTrace) {
        for (StackTraceElement stackTraceElement : stackTrace) {
```

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,280 @@
+public abstract class CrashReportAnalyser {
+
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final Map<String, IModInfo> basePackageModCache = new HashMap<>();
```
Uppercase constant names please, like LOGGER is

thedead2 committed:
Changed formatting.

thedead2 commented:
Changed the formatting

LexManos commented:
Interesting concept, I am however weary of putting anything this complex into the crash report system. We wouldn't want the crash report itself to crash. So make sure that anything you add fails gracefully just in case.

And paint is correct, we don't want to flat out blame explicit mods, so be sure that this is just listing things as "potential" causes. 

Aside from that, there are a few code concerns, namely your build string concatenation, could be made a lot cleaner using varargs. Can also be a lot more performant by making it build the string lazily. 

Also, your whole system of trying to find which packages belong to what mods can probably be replaced by using Module.getPackages().

thedead2 commented:
Of course crashing the crash report would be quite devastating. I'll add another try-catch block to the appendSuspectedMods() method just in case there is an exception while appending to the crash report. The analyseCrashReport() method already has a try-catch block as I saw more potential to crash there.

It is not intended to just blame mods to be the crash reason. That's why I called the section "suspected mods" and not "responsible mods" but I can change it to "potential causes" just in case there are misconceptions about that.

I thought about using Module.getPackages() but this returns all packages of the module. Where as caching all packages of a module to the corresponding modfile would be more accurate when trying to identify potential mods that caused a crash, it would certainly lead to a lot of entries inside the HashMap which might unnecessarily clutter the memory. I thought using just the base package names would be a good compromise for that.

I'm not quite sure how the use of varargs would make the code cleaner, isn't it irrelevant whether I use a String varargs or a String Array as the parameter for the findCommonClassPath() method or am I missing something.
How would you go about building the string lazily? Just use a StringBuilder to append the chars or is there a better way?

LexManos commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,280 @@
+                            "\n\t\tTarget" + (targetClasses.size() == 1 ? ": " + targetClasses.get(0).replaceAll("/", ".") : "s: " + targetClasses.toString().replaceAll("/", ".")) +
+                            "\n\t\tat " + elementAsString,
+                    mixinClassName
+            );
```
`blameIfPresent(mixinClassName, () -> blah)`
Making it a supplier, allows you to lazily resolve the string, which means that in the cases where there is no matching base package, you don't compute the big string for nothing.

LexManos commented:
> I thought about using Module.getPackages() but this returns all packages of the module. Where as caching all packages of a module to the corresponding modfile would be more accurate when trying to identify potential mods that caused a crash, it would certainly lead to a lot of entries inside the HashMap which might unnecessarily clutter the memory. I thought using just the base package names would be a good compromise for that.

Problem is what is the common package.. that common package can be shared by a ton of mods.
If you're worried about the memory, then you could lazily build the map so you only waste memory in the case of a crash.

LexManos commented on patches/minecraft/net/minecraft/CrashReport.java.patch:
```diff
@@ -51,3 +52,12 @@
+          stringbuilder.append("-");
+       }
+-
++      net.minecraftforge.logging.CrashReportAnalyser.analyseCrashReport(this.exception, this.uncategorizedStackTrace);
```
Why is this hooked in on this line? Honestly, looing at it, It looks like you should just make the CrashReportAnalyser class have a single entry point so that there is only a one line hook needed in this class. And we don't have to deal with the caching hooks elseware.

thedead2 commented:
Ok I just realised that I focused on the wrong method. When you were talking about varargs and lazy string building I only looked for the findCommonClassPath() method. For the blameIfPresent() method it makes total sense. I now used a String varargs for the blameIfPresent() method, concatenating the position string only if it's needed in the buildSuspectedModsSection() method from the array.

You're right, the hook for the analyseCrashReport() method is not needed. I changed it so there is only one hook inside the CrashReport class. I also changed the caching to only run on a crash and use all package names for caching. I'm however still a bit concerned about performance. It took 0.2 s to cache all packages of minecraft and forge. Of course they both have a lot of packages but there are mods out there that also have a lot of packages. If someone has a lot of mods installed, e.g. huge modpacks it would take quite a time to cache the package names. 
I experimented with a CompletableFuture to run the caching async on mod loading, but this would mean no lazy map building to save memory. I'm not quite sure which is more important, a fast performance or memory usage.
What's your opinion about that?

thedead2 committed:
Changed methods to use all package names for caching and changed when caching occurs

PaintNinja commented:
Run the `updateLicenses` Gradle task to fix the PR checks task failing

thedead2 committed:
Updated licenses...

PaintNinja reviewed (approved):
Has some minor issues like unnecessarily moving around imports and some places still using incorrect spacing (e.g. CrashReportAnalyser ln 72, 94 134...) but overall I really like the idea and the implementation looks alright.

thedead2 committed:
Changed the formatting again...

thedead2 commented:
I'm sorry for the formatting, changed this again. Hopefully I haven't missed anything. Would be easier if the link for the IDE formatter would still be valid. The moving around of imports happened through my IDE as I used the optimise imports action by habit, sorry for that.

PaintNinja commented:
No problem, it's a minor thing so not a big deal. :) Auto-formatting at build-time is on the todo list to eventually be sorted out.

Couple more nitpicks:
- You missed one import move (check the files changed tab)
- Rename "Issues URL" in the log to "Issue tracker URL" to line up with the mods.toml entry (makes it more obvious to devs on how to get theirs to show up there and isn't as specific to GitHub)

thedead2 committed:
Changed the imports order and changed Issues URL to Issue tracker URL...

thedead2 commented:
Fixed the unnecessarily moving around of imports

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,233 @@
+
+public abstract class CrashReportAnalyser {
+
+    private static final Logger LOGGER = LogManager.getLogger();
```
Please use Mojang's `LogUtils.getLogger();`

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,233 @@
+import java.util.*;
+
+
+public abstract class CrashReportAnalyser {
```
I suggest having a `final` class with an empty private constructor to avoid confusion.

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,233 @@
+        
+        return stringBuilder.toString();
+    }
+
+
+    public static Map<String, IModInfo> getPackageModCache() {
+        return PACKAGE_MOD_CACHE;
+    }
+
+    public static Map<IModInfo, String[]> getSuspectedMods() {
+        return SUSPECTED_MODS;
+    }
```
```suggestion
```
Unused public API from the old implementation can be removed

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,233 @@
+     * It checks the stacktrace for any occurrence of the package names cached in the {@link #PACKAGE_MOD_CACHE},
+     * including mixin classes.
+     * **/
+    public static void analyseCrashReport(Throwable throwable, StackTraceElement[] uncategorizedStackTrace) {
```
```suggestion
    private static void analyseCrashReport(Throwable throwable, StackTraceElement[] uncategorizedStackTrace) {
```

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,233 @@
+    /**
+     * Iterates over all loaded mods, resolving and caching their package names with the corresponding {@link IModInfo}.
+     * **/
+    public static void cacheModList() {
```
```suggestion
    private static void cacheModList() {
```

PaintNinja commented on src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java:
```diff
@@ -0,0 +1,233 @@
+    private static MixinMerged findMixinMerged(StackTraceElement element) {
+        try {
+            Class<?> clazz = Class.forName(element.getClassName());
+            Method[] methods = clazz.getDeclaredMethods();
+            for (Method method : methods) {
+                if (method.getName().equals(element.getMethodName())) {
```
Out of interest, what's the difference between this and using `clazz.getDeclaredMethod(element.getMethodName())`?

thedead2 replied:
The difference is that we don't know the params of the method and `clazz.getDeclaredMethod()` requires to pass in the classes of the parameters of the method, hence we can only check for the method's name.

PaintNinja reviewed (changes requested):
I've [followed up on this on the Discord server](https://discord.com/channels/1129059589325852724/1129095235889270844/1160314155463618570) and tried to commit the requested changes myself to save you some time, however I unfortunately don't seem to have permission to do so.

Please make some of the requested changes (particularly the public->private for a couple of methods and removing the two unused public getters) and it should be ready to merge. Once merged I'll look into backports as this'll be great to have for helping users troubleshoot things :)

thedead2 commented:
Hi thanks for the reply I wasn't that active on discord the last few days so I missed that. I now removed the getters. I initially left those there in case someone wants to use them e.g. mods like NotEnoughCrashes or similar.
Backports would be a great idea. To reply to Lex's suggestion to make this class a standalone mod, I think it wouldn't make sense to use this as a standalone mod. There are mods out there that have specialised in this kind of thing including preventing crashes of minecraft. But there are probably people out there that don't want to use the whole functionality of those mods and only would like to have the identification feature. Furthermore, I think not all of us want to use the aforementioned mods in a dev environment. Using them in such an environment would mean unnecessary dependencies and as far as maintenance is concerned, due to the fact that it only hooks into one line of code of minecraft, there shouldn't be a lot of maintenance required, despite maybe code improvements. 
And the last point I have to add is that mods may not be updated to newer versions of minecraft forge leading to missing features but with a thing implemented in forge itself it is certain to be updated to newer versions.

thedead2 committed:
Implemented the requested changes by PaintNinja

LexManos commented:
This is a good enough effort that should work until I get around to re-writing the stack trace/error message handler to make it less spammy/crazy.
